[azeventhubs] stress: $cbs link opened more than once

The $cbs link is used for authentication. It's an exclusive link - you can only have one of these open for your connection.

In our code we make this happen using a mutex, which protects the code that actually creates (and disposes) the $cbs link. I recently got an error in one of our stress tests where this somehow did NOT work, but it's unclear what led to that:

> 

There were some small windows where it seemed like we might leak a link on certain kinds of errors. Those are being addressed in PR#19326, but we'll need to keep an eye out and see if the error comes back. For now, I'm also making it so all the tests are in verbose mode in some runs.

Here are some log events. Note that we went 172 rounds (receiving 1,000,000 events each time) before it occurred.

1000000 events received
2022/10/29 11:08:27 [r:172/9223372036854775807,p:0] All messages received (1000000/1000000)
2022/10/29 11:08:27 [r:172/9223372036854775807,p:0] Done receiving messages from partition
2022/10/29 11:08:32 [r:173/9223372036854775807,p:0] Starting to receive messages from partition
2022/10/29 11:08:32 [r:173/9223372036854775807,p:0] Done receiving messages from partition
panic: received 0/1000000, but then got err: *Error{Condition: amqp:not-allowed, Description: A link to connection '2033874' $cbs node has already been opened., Info: map[]}

--- a/sdk/messaging/azeventhubs/CHANGELOG.md
+++ b/sdk/messaging/azeventhubs/CHANGELOG.md
@@ -1,14 +1,10 @@
 # Release History
 
-## 0.2.1 (Unreleased)
-
-### Features Added
-
-### Breaking Changes
+## 0.2.1 (2022-11-08)
 
 ### Bugs Fixed
 
-### Other Changes
+- $cbs link is properly closed, even on cancellation (#19492)
 
 ## 0.2.0 (2022-10-17)
 

--- a/sdk/messaging/azeventhubs/internal/eh/stress/Chart.yaml
+++ b/sdk/messaging/azeventhubs/internal/eh/stress/Chart.yaml
@@ -8,5 +8,5 @@ annotations:
   namespace: 'go'
 dependencies:
 - name: stress-test-addons
-  version: ~2.0.0
+  version: ~0.2.0
   repository: https://stresstestcharts.blob.core.windows.net/helm/

--- a/sdk/messaging/azeventhubs/internal/eh/stress/Chart.yaml
+++ b/sdk/messaging/azeventhubs/internal/eh/stress/Chart.yaml
@@ -8,5 +8,5 @@ annotations:
   namespace: 'go'
 dependencies:
 - name: stress-test-addons
-  version: ~0.1.20
+  version: ~2.0.0
   repository: https://stresstestcharts.blob.core.windows.net/helm/

--- a/sdk/messaging/azeventhubs/internal/links.go
+++ b/sdk/messaging/azeventhubs/internal/links.go
@@ -310,7 +310,11 @@ func (l *Links[LinkT]) Close(ctx context.Context) error {
 	return l.closeLinks(ctx, true)
 }
 
-func (l *Links[LinkT]) closeLinks(ctx context.Context, permanent bool) error {
+func (l *Links[LinkT]) closeLinks(_ context.Context, permanent bool) error {
+	// we're finding, in practice, that allowing cancellations when cleaning up state
+	// just results in inconsistencies. We'll cut cancellation off here for now.
+	ctx := context.Background()
+
 	if err := l.closeManagementLink(ctx); err != nil {
 		azlog.Writef(exported.EventConn, "Error while cleaning up management link while doing connection recovery: %s", err.Error())
 	}

--- a/sdk/messaging/azeventhubs/internal/rpc.go
+++ b/sdk/messaging/azeventhubs/internal/rpc.go
@@ -100,13 +100,10 @@ type RPCLinkArgs struct {
 	LogEvent azlog.Event
 }
 
-func closeOrLog(name string, limit time.Duration, closeable interface {
+func closeOrLog(name string, closeable interface {
 	Close(ctx context.Context) error
 }) {
-	ctx, cancel := context.WithTimeout(context.Background(), limit)
-	defer cancel()
-
-	if err := closeable.Close(ctx); err != nil {
+	if err := closeable.Close(context.Background()); err != nil {
 		log.Writef(exported.EventAuth, "Failed closing %s for RPC Link: %s", name, err.Error())
 	}
 }
@@ -121,7 +118,7 @@ func NewRPCLink(ctx context.Context, args RPCLinkArgs) (*rpcLink, error) {
 
 	linkID, err := uuid.New()
 	if err != nil {
-		closeOrLog("session", 5*time.Second, session)
+		closeOrLog("session", session)
 		return nil, err
 	}
 
@@ -143,7 +140,7 @@ func NewRPCLink(ctx context.Context, args RPCLinkArgs) (*rpcLink, error) {
 		nil,
 	)
 	if err != nil {
-		closeOrLog("session", 5*time.Second, session)
+		closeOrLog("session", session)
 		return nil, err
 	}
 
@@ -164,8 +161,8 @@ func NewRPCLink(ctx context.Context, args RPCLinkArgs) (*rpcLink, error) {
 
 	receiver, err := session.NewReceiver(ctx, args.Address, receiverOpts)
 	if err != nil {
-		closeOrLog("sender", 5*time.Second, sender)
-		closeOrLog("session", 5*time.Second, session)
+		closeOrLog("sender", sender)
+		closeOrLog("session", session)
 		return nil, err
 	}
 
@@ -334,7 +331,11 @@ func (l *rpcLink) RPC(ctx context.Context, msg *amqp.Message) (*RPCResponse, err
 }
 
 // Close the link receiver, sender and session
-func (l *rpcLink) Close(ctx context.Context) error {
+func (l *rpcLink) Close(_ context.Context) error {
+	// we're finding, in practice, that allowing cancellations when cleaning up state
+	// just results in inconsistencies. We'll cut cancellation off here for now.
+	ctx := context.Background()
+
 	l.rpcLinkCtxCancel()
 
 	if err := l.closeReceiver(ctx); err != nil {

--- a/sdk/messaging/azservicebus/CHANGELOG.md
+++ b/sdk/messaging/azservicebus/CHANGELOG.md
@@ -1,14 +1,10 @@
 # Release History
 
-## 1.1.2 (Unreleased)
-
-### Features Added
-
-### Breaking Changes
+## 1.1.2 (2022-11-08)
 
 ### Bugs Fixed
 
-### Other Changes
+- $cbs link is properly closed, even on cancellation (#19492)
 
 ## 1.1.1 (2022-10-11)
 

--- a/sdk/messaging/azservicebus/internal/amqpLinks.go
+++ b/sdk/messaging/azservicebus/internal/amqpLinks.go
@@ -468,7 +468,9 @@ func (l *AMQPLinksImpl) initWithoutLocking(ctx context.Context) error {
 
 // close closes the link.
 // NOTE: No locking is done in this function, call `Close` if you require locking.
-func (l *AMQPLinksImpl) closeWithoutLocking(ctx context.Context, permanent bool) error {
+func (l *AMQPLinksImpl) closeWithoutLocking(_ context.Context, permanent bool) error {
+	ctx := context.Background()
+
 	if l.closedPermanently {
 		return nil
 	}

--- a/sdk/messaging/azservicebus/internal/rpc.go
+++ b/sdk/messaging/azservicebus/internal/rpc.go
@@ -11,9 +11,11 @@ import (
 	"sync"
 	"time"
 
+	"github.com/Azure/azure-sdk-for-go/sdk/internal/log"
 	azlog "github.com/Azure/azure-sdk-for-go/sdk/internal/log"
 	"github.com/Azure/azure-sdk-for-go/sdk/internal/uuid"
 	"github.com/Azure/azure-sdk-for-go/sdk/messaging/azservicebus/internal/amqpwrap"
+	"github.com/Azure/azure-sdk-for-go/sdk/messaging/azservicebus/internal/exported"
 	"github.com/Azure/azure-sdk-for-go/sdk/messaging/azservicebus/internal/go-amqp"
 )
 
@@ -98,6 +100,14 @@ type RPCLinkArgs struct {
 	LogEvent azlog.Event
 }
 
+func closeOrLog(name string, closeable interface {
+	Close(ctx context.Context) error
+}) {
+	if err := closeable.Close(context.Background()); err != nil {
+		log.Writef(exported.EventAuth, "Failed closing %s for RPC Link: %s", name, err.Error())
+	}
+}
+
 // NewRPCLink will build a new request response link
 func NewRPCLink(ctx context.Context, args RPCLinkArgs) (*rpcLink, error) {
 	session, err := args.Client.NewSession(ctx, nil)
@@ -108,6 +118,7 @@ func NewRPCLink(ctx context.Context, args RPCLinkArgs) (*rpcLink, error) {
 
 	linkID, err := uuid.New()
 	if err != nil {
+		closeOrLog("session", session)
 		return nil, err
 	}
 
@@ -129,6 +140,7 @@ func NewRPCLink(ctx context.Context, args RPCLinkArgs) (*rpcLink, error) {
 		nil,
 	)
 	if err != nil {
+		closeOrLog("session", session)
 		return nil, err
 	}
 
@@ -149,11 +161,8 @@ func NewRPCLink(ctx context.Context, args RPCLinkArgs) (*rpcLink, error) {
 
 	receiver, err := session.NewReceiver(ctx, args.Address, receiverOpts)
 	if err != nil {
-		// make sure we close the sender
-		clsCtx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
-		defer cancel()
-
-		_ = sender.Close(clsCtx)
+		closeOrLog("sender", sender)
+		closeOrLog("session", session)
 		return nil, err
 	}
 
@@ -322,7 +331,11 @@ func (l *rpcLink) RPC(ctx context.Context, msg *amqp.Message) (*RPCResponse, err
 }
 
 // Close the link receiver, sender and session
-func (l *rpcLink) Close(ctx context.Context) error {
+func (l *rpcLink) Close(_ context.Context) error {
+	// we're finding, in practice, that allowing cancellations when cleaning up state
+	// just results in inconsistencies. We'll cut cancellation off here for now.
+	ctx := context.Background()
+
 	l.rpcLinkCtxCancel()
 
 	if err := l.closeReceiver(ctx); err != nil {

--- a/sdk/messaging/azservicebus/internal/stress/Chart.yaml
+++ b/sdk/messaging/azservicebus/internal/stress/Chart.yaml
@@ -8,5 +8,5 @@ annotations:
   namespace: 'go'
 dependencies:
 - name: stress-test-addons
-  version: ~2.0.0
+  version: ~0.2.0
   repository: https://stresstestcharts.blob.core.windows.net/helm/

--- a/sdk/messaging/azservicebus/internal/stress/Chart.yaml
+++ b/sdk/messaging/azservicebus/internal/stress/Chart.yaml
@@ -8,5 +8,5 @@ annotations:
   namespace: 'go'
 dependencies:
 - name: stress-test-addons
-  version: ~0.1.20
+  version: ~2.0.0
   repository: https://stresstestcharts.blob.core.windows.net/helm/

